File: teamapps-client/teamapps-client-core-components/src/main/ts/component/Notification.ts
/*-
 * ========================LICENSE_START=================================
 * TeamApps
 * ---
 * Copyright (C) 2014 - 2022 TeamApps.org
 * ---
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =========================LICENSE_END==================================
 */
import {
	DtoNotification,
	DtoNotification_ClosedEvent,
	DtoNotification_OpenedEvent,
	DtoNotificationCommandHandler,
	DtoNotificationEventSource,
	NotificationPosition
} from "../generated";
import {
	AbstractLegacyComponent,
	Component, EntranceAnimation, ExitAnimation,
	executeWhenFirstDisplayed,
	parseHtml,
	ServerObjectChannel,
	TeamAppsEvent, animateCSS
} from "projector-client-object-api";
import {createUiSpacingValueCssString} from "projector-client-object-api";
import {ProgressBar} from "../micro-components/ProgressBar";

const containersByPosition: {
	[NotificationPosition.TOP_LEFT]: HTMLElement,
	[NotificationPosition.TOP_CENTER]: HTMLElement,
	[NotificationPosition.TOP_RIGHT]: HTMLElement,
	[NotificationPosition.BOTTOM_LEFT]: HTMLElement,
	[NotificationPosition.BOTTOM_CENTER]: HTMLElement,
	[NotificationPosition.BOTTOM_RIGHT]: HTMLElement
} = {
	[NotificationPosition.TOP_LEFT]: parseHtml(`<div class="Notification-container top-left"></div>`),
	[NotificationPosition.TOP_CENTER]: parseHtml(`<div class="Notification-container top-center"></div>`),
	[NotificationPosition.TOP_RIGHT]: parseHtml(`<div class="Notification-container top-right"></div>`),
	[NotificationPosition.BOTTOM_LEFT]: parseHtml(`<div class="Notification-container bottom-left"></div>`),
	[NotificationPosition.BOTTOM_CENTER]: parseHtml(`<div class="Notification-container bottom-center"></div>`),
	[NotificationPosition.BOTTOM_RIGHT]: parseHtml(`<div class="Notification-container bottom-right"></div>`)
};

let notifications: {
	notification: Notification;
	position: NotificationPosition;
	$wrapper: HTMLElement;
}[] = [];

function getNotificationsByPosition(position: NotificationPosition) {
	return notifications.filter(n => n.position == position);
}

let updateContainerVisibilities = function () {
	[NotificationPosition.TOP_LEFT,
		NotificationPosition.TOP_CENTER,
		NotificationPosition.TOP_RIGHT,
		NotificationPosition.BOTTOM_LEFT,
		NotificationPosition.BOTTOM_CENTER,
		NotificationPosition.BOTTOM_RIGHT].forEach(pos => {
		let hasNotifications = getNotificationsByPosition(pos).length > 0;
		if (hasNotifications && containersByPosition[pos].parentNode !== document.body) {
			document.body.appendChild(containersByPosition[pos]);
		} else if (!hasNotifications) {
			containersByPosition[pos].remove();
		}
	});
};

export function showNotification(notification: Notification, position: NotificationPosition, entranceAnimation: EntranceAnimation, exitAnimation: ExitAnimation) {
	let notif = notifications.filter(n => n.notification == notification)[0];

	if (notif == null || notif.position != position) {
		if (notif == null) {
			let $wrapper = parseHtml(`<div class="notification-wrapper"></div>`);
			$wrapper.appendChild(notification.getMainElement());
			notif = {notification, position, $wrapper};
			notifications.push(notif)
		} else {
			notif.position = position;
		}
		notif.$wrapper.style.height = null;
		notif.$wrapper.style.marginBottom = null;
		notif.$wrapper.style.zIndex = null;
		containersByPosition[position].appendChild(notif.$wrapper);

		updateContainerVisibilities();

		animateCSS(notification.getMainElement(), entranceAnimation, 700);

		let closeListener = () => {
			notification.onClosedAnyWay.removeListener(closeListener);
			notif.$wrapper.style.height = `${notif.$wrapper.offsetHeight}px`;
			notif.$wrapper.offsetHeight; // make sure the style above is applied so we get a transition!
			notif.$wrapper.style.height = "0px";
			notif.$wrapper.style.marginBottom = "0px";
			notif.$wrapper.style.zIndex = "0";

			animateCSS(notification.getMainElement(), exitAnimation, 700, () => {
				notif.$wrapper.remove();
				updateContainerVisibilities();
			});

			notifications = notifications.filter(n => n.notification !== notification);
		};
		notification.onClosedAnyWay.addListener(closeListener);

		notification.onOpened.fire({});
	}

	setTimeout(() => notification.startCloseTimeout());
}

export class Notification extends AbstractLegacyComponent<DtoNotification> implements DtoNotificationCommandHandler, DtoNotificationEventSource {

	public readonly onOpened: TeamAppsEvent<DtoNotification_OpenedEvent> = new TeamAppsEvent();
	public readonly onClosed: TeamAppsEvent<DtoNotification_ClosedEvent> = new TeamAppsEvent();
	public readonly onClosedAnyWay: TeamAppsEvent<void> = new TeamAppsEvent();

	private $main: HTMLElement;
	private $contentContainer: HTMLElement;
	private $progressBarContainer: HTMLElement;
	private progressBar: ProgressBar;

	constructor(config: DtoNotification, serverObjectChannel: ServerObjectChannel) {
		super(config, serverObjectChannel);

		this.$main = parseHtml(`<div class="Notification">
	<div class="close-button"></div>
	<div class="content-container"></div>
	<div class="progress-container"></div>
</div>`);
		this.$contentContainer = this.$main.querySelector(":scope > .content-container");
		this.$progressBarContainer = this.$main.querySelector(":scope > .progress-container");
		this.$main.querySelector(":scope > .close-button").addEventListener("mousedown", () => {
			this.close();
			this.onClosed.fire({byUser: true});
		});
		this.update(config);
	}

	public update(config: DtoNotification) {
		this.config = config;
		this.$main.style.backgroundColor = config.backgroundColor;
		// this.$main.style.borderColor = createUiColorCssString(config.borderColor, "#00000022");
		this.$contentContainer.style.padding = createUiSpacingValueCssString(config.padding);
		this.$main.classList.toggle("dismissible", config.dismissible);
		this.$main.classList.toggle("show-progress", config.progressBarVisible && config.displayTimeInMillis > 0);

		if (config.progressBarVisible && this.progressBar == null) {
			this.progressBar = new ProgressBar(0, {height: 5, transitionTime: config.displayTimeInMillis});
			this.$progressBarContainer.appendChild(this.progressBar.getMainDomElement());
		} else if (!config.progressBarVisible && this.progressBar != null) {
			this.progressBar.getMainDomElement().remove();
			this.progressBar = null;
		}

		if (this.$contentContainer.firstChild !== (config.content && (config.content as Component).getMainElement())) {
			this.$contentContainer.innerHTML = '';
			if (config.content != null) {
				this.$contentContainer.appendChild((config.content as Component).getMainElement());
			}
		}
	}

	public static showNotification(notification: Notification, position: NotificationPosition, entranceAnimation: EntranceAnimation, exitAnimation: ExitAnimation) {
		showNotification(notification, position, entranceAnimation, exitAnimation);
	}

	private closeTimeout: number;

	@executeWhenFirstDisplayed(true)
	public startCloseTimeout() {
		if (this.progressBar != null) {
			this.progressBar.setProgress(1); // mind the css transition!
		}
		if (this.closeTimeout != null) {
			window.clearTimeout(this.closeTimeout);
		}
		if (this.config.displayTimeInMillis > 0) {
			this.closeTimeout = window.setTimeout(() => {
				this.close();
				this.onClosed.fire({byUser: false});
				this.onClosedAnyWay.fire();
			}, this.config.displayTimeInMillis);
		}
	}

	doGetMainElement(): HTMLElement {
		return this.$main;
	}

	close(): void {
		this.onClosedAnyWay.fire();
	}

}

File: teamapps-client/projector-notification/src/main/java/org/teamapps/projector/notification/Notification.java
/*-
 * ========================LICENSE_START=================================
 * TeamApps
 * ---
 * Copyright (C) 2014 - 2022 TeamApps.org
 * ---
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =========================LICENSE_END==================================
 */
package org.teamapps.projector.notification;

import org.teamapps.common.format.Color;
import org.teamapps.icons.Icon;
import org.teamapps.projector.animation.EntranceAnimation;
import org.teamapps.projector.animation.ExitAnimation;
import org.teamapps.projector.annotation.ClientObjectLibrary;
import org.teamapps.projector.component.AbstractComponent;
import org.teamapps.projector.component.Component;
import org.teamapps.projector.component.essential.field.TemplateField;
import org.teamapps.projector.component.notification.DtoNotification;
import org.teamapps.projector.component.notification.DtoNotificationClientObjectChannel;
import org.teamapps.projector.component.notification.DtoNotificationEventHandler;
import org.teamapps.projector.event.ProjectorEvent;
import org.teamapps.projector.format.Spacing;
import org.teamapps.projector.template.grid.basetemplates.BaseTemplateRecord;
import org.teamapps.projector.template.grid.basetemplates.BaseTemplates;

import java.time.Duration;

@ClientObjectLibrary(value = NotificationLibrary.class)
public class Notification extends AbstractComponent implements DtoNotificationEventHandler {

	private final DtoNotificationClientObjectChannel clientObjectChannel = new DtoNotificationClientObjectChannel(getClientObjectChannel());

	public final ProjectorEvent<Void> onOpened = new ProjectorEvent<>(clientObjectChannel::toggleOpenedEvent);
	public final ProjectorEvent<Boolean> onClosed = new ProjectorEvent<>(clientObjectChannel::toggleClosedEvent);

	private boolean showing;

	private Color backgroundColor = null;
	private Spacing padding = null;
	private boolean dismissible = true;
	private boolean progressBarEnabled = true;

	private Component content;

	public Notification() {
	}

	public Notification(Component content) {
		this.content = content;
	}

	public static Notification createWithIconAndCaption(Icon<?, ?> icon, String text) {
		TemplateField<BaseTemplateRecord<Void>> templateField = new TemplateField<>(BaseTemplates.NOTIFICATION_ICON_CAPTION);
		templateField.setValue(new BaseTemplateRecord<>(icon, text));
		Notification notification = new Notification();
		notification.setContent(templateField);
		return notification;
	}

	public static Notification createWithIconCaptionDescription(Icon<?, ?> icon, String caption, String description) {
		TemplateField<BaseTemplateRecord<Void>> templateField = new TemplateField<>(BaseTemplates.NOTIFICATION_ICON_CAPTION_DESCRIPTION);
		templateField.setValue(new BaseTemplateRecord<>(icon, caption, description));
		Notification notification = new Notification();
		notification.setContent(templateField);
		return notification;
	}

	public DtoNotification createConfig() {
		DtoNotification ui = new DtoNotification();
		mapAbstractConfigProperties(ui);
		ui.setBackgroundColor(backgroundColor != null ? backgroundColor.toHtmlColorString() : null);
		ui.setPadding(padding != null ? padding.createUiSpacing() : null);
		ui.setDismissible(dismissible);
		ui.setProgressBarVisible(progressBarEnabled);
		ui.setContent(content != null ? content : null);
		return ui;
	}

	@Override
	public void handleOpened() {
		this.showing = true;
		onOpened.fire(null);
	}

	@Override
	public void handleClosed(boolean byUser) {
		this.showing = false;
		onClosed.fire(byUser);
	}

	public void close() {
		clientObjectChannel.close();
	}

	public Color getBackgroundColor() {
		return backgroundColor;
	}

	public Notification setBackgroundColor(Color backgroundColor) {
		this.backgroundColor = backgroundColor;
		clientObjectChannel.update(createConfig());
		return this;
	}

	public Spacing getPadding() {
		return padding;
	}

	public Notification setPadding(Spacing padding) {
		this.padding = padding;
		clientObjectChannel.update(createConfig());
		return this;
	}

	public boolean isDismissible() {
		return dismissible;
	}

	public Notification setDismissible(boolean dismissible) {
		this.dismissible = dismissible;
		clientObjectChannel.update(createConfig());
		return this;
	}

	public boolean isProgressBarEnabled() {
		return progressBarEnabled;
	}

	public Notification setProgressBarEnabled(boolean progressBarEnabled) {
		this.progressBarEnabled = progressBarEnabled;
		clientObjectChannel.update(createConfig());
		return this;
	}

	public Component getContent() {
		return content;
	}

	public Notification setContent(Component content) {
		this.content = content;
		clientObjectChannel.update(createConfig());
		return this;
	}

	public void showTillDismissed() {
		show(Duration.ZERO);
	}

	public void show(Duration timeout) {
		show(timeout, NotificationPosition.TOP_RIGHT);
	}

	public void show(Duration timeout, NotificationPosition position) {
		show(timeout, position, EntranceAnimation.SLIDE_IN_LEFT, ExitAnimation.SLIDE_OUT_DOWN);
	}

	public void show(Duration timeout, NotificationPosition position, EntranceAnimation entranceAnimation, ExitAnimation exitAnimation) {
		clientObjectChannel.forceRender();
		clientObjectChannel.show(position, entranceAnimation, exitAnimation, timeout.toMillis());
	}

	public boolean isShowing() {
		return showing;
	}

}
